scripts/locust/service.py

```diff
@@ -40,120 +81,251 @@ def parallel_requests():
     group.join()
 
 
-class ServiceTaskSet(TaskSet):
-    """This is the default page so start here"""
+# These task-less parent classes define common methods involving the locust HTTP
+# client.
+class AzulTaskSet(SequentialTaskSet):
+    catalog: str = configured_catalog()
 
-    def on_start(self):
-        self.projects_page()
+    def endpoint(self, path: str, **query) -> str:
+        return furl(path=path, query=dict(query, catalog=self.catalog)).url
 
-    def projects_page(self):
-        with parallel_requests() as group:
-            group.spawn(lambda: self.client.get('/index/summary?filters=%7B%7D'))
-            group.spawn(lambda: self.client.get('/index/projects?filters=%7B%7D&size=15'))
 
-    @task
-    def filter_mtx_files(self):
+class BrowserTaskSet(AzulTaskSet):
+    browser_search_params = dict(size=15, sort='entryId', order='desc')
+
+    def browse_page(self,
+                    index_name: str,
+                    filters: JSON,
+                    **extra_index_params):
+        filters = json.dumps(filters)
         with parallel_requests() as group:
-            group.spawn(lambda: self.client.get('/index/summary?filters=%7B%22file%22%3A%7B%22fileFormat'
-                                                '%22%3A%7B%22is%22%3A%5B%22mtx%22%5D%7D%7D%7D'))
-            group.spawn(lambda: self.client.get('/index/projects?filters=%7B%22file%22%3A%7B%22fileFormat'
-                                                '%22%3A%7B%22is%22%3A%5B%22mtx%22%5D%7D%7D%7D'
-                                                '&size=15&sort=sampleId&order=desc'))
-
-    @task(3)
-    class FilesTaskSet(TaskSequence):
+            group.spawn(lambda: self.client.get(self.endpoint('/index/summary',
+                                                              filters=filters),
+                                                name='/index/summary'))
+            group.spawn(lambda: self.client.get(self.endpoint(f'/index/{index_name}',
+                                                              filters=filters,
+                                                              **extra_index_params),
+                                                name=f'/index/{index_name}'))
+
+
+class FileSelectionTaskSet(AzulTaskSet):
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.metadata_uuid = None
+        self.file_uuid = None
+
+    def choose_file(self, filters: JSON):
+        with self.client.get(self.endpoint('/index/files',
+                                           filters=json.dumps(filters),
+                                           size=15),
+                             name='/index/files') as response:
+            hit = random.choice(response.json()['hits'])
+            self.metadata_uuid = hit['entryId']
+            self.file_uuid = one(hit['files'])['uuid']
+
+
+class FileFetchTaskSet(AzulTaskSet):
+    @dataclass
```

Comment:
We are moving to `@attr.s` with a preference to `frozen=True` and `kwonly=True`.